rmw_connextdds_common/src/ndds/dds_api_ndds.cpp

```diff
@@ -197,25 +200,119 @@ rmw_connextdds_initialize_participant_qos_impl(
+  if (!ctx->domain_tag) {
+    const auto pid = rcutils_get_pid();
+    ctx->domain_tag = DDS_String_alloc(18 + log10(pid));
+    sprintf(ctx->domain_tag, "ros_discovery_off_%d", pid);
+  }
 
   switch (ctx->participant_qos_override_policy) {
     case rmw_context_impl_t::participant_qos_override_policy_t::All:
     case rmw_context_impl_t::participant_qos_override_policy_t::Basic:
       {
         // Parse and apply QoS parameters derived from ROS 2 configuration options.
+        // Reference link for properties:
+        // https://community.rti.com/static/documentation/connext-dds/6.1.1/doc/manuals/connext_dds_professional/properties_reference/index.html
+        if (ctx->discovery_options) {
+          const auto range = ctx->discovery_options->automatic_discovery_range;
+          switch (range) {
+            case RMW_AUTOMATIC_DISCOVERY_RANGE_SUBNET:
+              /* No action needed. This is the default discovery behavior for DDS */
+              break;
+            default:
+              RMW_CONNEXT_LOG_WARNING_A(
+                "Unknown value provided for automatic discovery range: %i",
+                ctx->discovery_options->automatic_discovery_range);
+              /* Fall back to the default behavior */
+              [[clang::fallthrough]];
+              /* [[fallthrough]]; // Uncomment this when migrating to C++17 */
+            case RMW_AUTOMATIC_DISCOVERY_RANGE_DEFAULT:
+              /* Same behavior as LOCALHOST */
+              [[clang::fallthrough]];
+              /* [[fallthrough]]; // Uncomment this when migrating to C++17 */
+            case RMW_AUTOMATIC_DISCOVERY_RANGE_LOCALHOST:
+              /* Same interface settings as OFF */
+              [[clang::fallthrough]];
+              /* [[fallthrough]]; // Uncomment this when migrating to C++17 */
+            case RMW_AUTOMATIC_DISCOVERY_RANGE_OFF:
+              /* Note: We allow the LOCALHOST interface for the OFF range
+                 because if we leave this property completely blank then it
+                 has the opposite effect and allows all interfaces to be used.
+                 Allowing only LOCALHOST at least minimizes the unnecessary
+                 discovery traffic and prevents discovery with other host
+                 machines, while the domain_tag protects against same-host
+                 connections. */
+              if (DDS_RETCODE_OK != DDS_PropertyQosPolicyHelper_assert_property(
+                &dp_qos->property,
+                "dds.transport.UDPv4.builtin.parent.allow_multicast_interfaces_list",
+                RMW_CONNEXT_LOCALHOST_ONLY_ADDRESS,
+                DDS_BOOLEAN_FALSE /* propagate */))
+              {
+                RMW_CONNEXT_LOG_ERROR_A_SET(
+                  "failed to assert property on participant: %s",
+                  "dds.transport.UDPv4.builtin.parent.allow_multicast_interfaces_list")
+                return RMW_RET_ERROR;
+              }
+          }
```

Comment:
I'm not familiar with this idiom, and I'm not sure how C++17 would change this block, but I definitely dislike this "fall through" block.

It can be simplified by using a boolean flag that is set separately in the block for the valid enums, and in the `default` one (which should be moved after)

```suggestion
          bool set_multicast_peers = false;
          switch (range) {
            case RMW_AUTOMATIC_DISCOVERY_RANGE_SUBNET:
              /* No action needed. This is the default discovery behavior for DDS */
              break;
            case RMW_AUTOMATIC_DISCOVERY_RANGE_DEFAULT:
            case RMW_AUTOMATIC_DISCOVERY_RANGE_LOCALHOST:
            case RMW_AUTOMATIC_DISCOVERY_RANGE_OFF:
              set_multicast_peers = false;
              break; 
            default:
              RMW_CONNEXT_LOG_WARNING_A(
                "Unknown value provided for automatic discovery range: %i",
                ctx->discovery_options->automatic_discovery_range);
              set_multicast_peers = false;
              break;
          }
          if (set_multicast_peers) {
            /* Note: We allow the LOCALHOST interface for the OFF range
               because if we leave this property completely blank then it
               has the opposite effect and allows all interfaces to be used.
               Allowing only LOCALHOST at least minimizes the unnecessary
               discovery traffic and prevents discovery with other host
               machines, while the domain_tag protects against same-host
               connections. */
            if (DDS_RETCODE_OK != DDS_PropertyQosPolicyHelper_assert_property(
                &dp_qos->property,
                "dds.transport.UDPv4.builtin.parent.allow_multicast_interfaces_list",
                RMW_CONNEXT_LOCALHOST_ONLY_ADDRESS,
                DDS_BOOLEAN_FALSE /* propagate */))
            {
              RMW_CONNEXT_LOG_ERROR_A_SET(
                "failed to assert property on participant: %s",
                "dds.transport.UDPv4.builtin.parent.allow_multicast_interfaces_list")
              return RMW_RET_ERROR;
            }
        }
```

Reply:
I didn't notice this comment until after I changed some things about the fallthrough stuff. But I attempted it in b4c9e9c.